src/main/java/server/ChatServer.java

```diff
@@ -8,118 +8,78 @@
 
 // Server class
 public class ChatServer {
-
     // Vector to store active clients
     static Vector<ClientHandler> ar = new Vector<>();
 
+    public static void enqueue(Runnable r) {
+        Thread t = new Thread(r);
+        t.start();
+    }
 
-    public static void main(String[] args) throws IOException {
-        // list of approved users
-        UserLogin userlogin = new UserLogin();
-        List<String> userList = userlogin.listOfUsers();
-
-        // server is listening on port 8000
-        ServerSocket ss = new ServerSocket(8000);
-        System.out.println("Server started");
-        Socket s;
-
-        // running infinite loop for getting
-        // client request
-        while (true) {
-            // Accept the incoming request
-            s = ss.accept();
-
-            System.out.println("New client request received : " + s);
-
-            // obtain input and output streams
-            DataInputStream dis = new DataInputStream(s.getInputStream());
-            PrintWriter dos = new PrintWriter(s.getOutputStream(),true);
-            StringTokenizer st = new StringTokenizer(dis.readLine(), "#");
-            String cmd = st.nextToken();
-
-            try {
-
-
-                if (cmd.equals("CONNECT")) {
-                    String username = st.nextToken();
-                    int i = 0;
-
-
-                    for (String users : userList) {
-
-
-                        if (username.equals(users)) {
-
-                            System.out.println("Creating a new handler for this client...");
-
-                            // Create a new handler object for handling this request.
-                            ClientHandler mtch = new ClientHandler(s, username, dis, dos);
-
-                            // Create a new Thread with this object.
-                            Thread t = new Thread(mtch);
-
-                            System.out.println("Adding this client to active client list");
-
-                            // add this client to active clients list
-                            ar.add(mtch);
-
-                            // start the thread.
-                            t.start();
-
-                            onlineMessage();
-
-                            break;
-
-                        }
-                        i++;
-
-                        if (i == userList.size()) {
-                            dos.println("CLOSE#2");
-                            s.close();
-                            break;
-                        }
-
-
-                    }
+    /**
+     * sendMessage attempts to send a message to the given destination.
+     */
+    public static void sendMessage(ClientHandler source, String destination, String message) {
```

Comment:
sendMessage is placed here because this function iterates over `ChatServer.ar`, and you really don't want to have a public value that changes referenced in other classes if you can help it.

This is still not fantastic because the value is static and mutable and this is almost certainly not what you want. This also maintains the circular dependency between ClientHandler and ChatServer, which, again, not good, but good enough for this use case.

This does mean that you can more easily refactor sending messages to be multi-thread safe by adding sycnhronisation primitives to protect `ar` (to prevent reading and modification at the same time).